tests/beacon/types/test_states.py

```diff
@@ -20,62 +20,39 @@
 )
 
 
-@pytest.fixture
-def empty_beacon_state():
-    return BeaconState(
-        slot=0,
-        genesis_time=0,
-        fork_data=ForkData(
-            pre_fork_version=0,
-            post_fork_version=0,
-            fork_slot=0,
-        ),
-        validator_registry=(),
-        validator_registry_latest_change_slot=10,
-        validator_registry_exit_count=10,
-        validator_registry_delta_chain_tip=b'\x55' * 32,
-        latest_randao_mixes=(),
-        latest_vdf_outputs=(),
-        shard_committees_at_slots=(),
-        persistent_committees=(),
-        persistent_committee_reassignments=(),
-        previous_justified_slot=0,
-        justified_slot=0,
-        justification_bitfield=0,
-        finalized_slot=0,
-        latest_crosslinks=(),
-        latest_block_roots=(),
-        latest_penalized_exit_balances=(),
-        latest_attestations=(),
-        batched_block_roots=(),
-        processed_pow_receipt_root=b'\x55' * 32,
-        candidate_pow_receipt_roots=(),
-    )
-
-
 def test_defaults(sample_beacon_state_params):
     state = BeaconState(**sample_beacon_state_params)
     assert state.validator_registry == sample_beacon_state_params['validator_registry']
     assert state.validator_registry_latest_change_slot == sample_beacon_state_params['validator_registry_latest_change_slot']  # noqa: E501
 
 
+def test_validator_registry_and_balances_length(sample_beacon_state_params, far_future_slot):
+    # When len(BeaconState.validator_registry) != len(BeaconState.validtor_balances)
+    with pytest.raises(ValueError):
```

Comment:
Is `ValidationError` a more appropriate error here?

Reply:
The _user_ should never be able to provide data/input that would cause this to trigger. A _developer_ could. Does that make it a `ValidationError`?

Reply:
@djrtwo yes, that's also what I thought! We should ensure `len(BeaconState.validator_registry) == len(BeaconState.validtor_balances)` during writing  `eth` APIs.

Reply:
@djrtwo 

- `ValidationError` **if** it is *possible* for this code path to be encountered under some form of live conditions like *somehow* ending up with an invalid state where the registry and balances get out of sync.
- `ValueError` for lower level errors, normally indicative of us having some kind of broken code which is passing in the wrong type of value or a malformed value.

Admittedly these two definitions are not fully independent.  Maybe rule of thumb is: 1) A user should *never* see a `ValueError`, nor should anything like the `ethereum/tests` end up triggering a `ValueError`.  `ValidationError` **may** occasionally show up for users and is likely a regular result for many of the `ethereum/tests` style tests which test edge cases.

Reply:
Hmmm. I'm still a bit unclear on this.
There is an invariant that and addition to or removal from one of these arrays should always require an symmetric addition/removal from the other array. All helper functions such as `add_pending_validator` should enforce this invariant.

The state would only be out of sync if one of these helper functions was broken or if a user decided to manipulate/create a state directly with mismatched arrays. Because of the assumed atomicity of operations affecting the length of these two arrays, the network should never be able to provide data that gets them out of sync. This is different than say, the network providing an invalid block that throws a `ValidationError` when processing.

I somewhat think `ValueError`, but I am still trying to grok the difference and hope for some more clarity :)

Reply:
@pipermerriam thanks for clarifying! FYI the reason why we put validator balances into a separate list in BeaconState (#1647) was clarified [here](https://github.com/ethereum/eth2.0-specs/issues/386#issuecomment-451033847) by Justin:
> It's an optimisation to minimise the amount of hashing when computing the beacon state root. The `validator_registry` (which contains most of the state) changes slowly and requires little hashing effort to maintain. The `validator_balances` is a smaller part of the state with much higher churn, hence why it is segregated.

So I do think this `ValueError` here fits "*for lower level errors, normally indicative of us having some kind of broken code which is passing in the wrong type of value or a malformed value*" since it's for avoiding reckless coding.

Reply:
:+1: seems like this could fit either case and that's expected since the two cases do have some conceptual overlap.  No strong opinion of what should be done if anything here.